Images Tab gives Error: HTTP error! status: 504

Happy new year!

After syncing, I am getting Error: HTTP error! status: 504 on images tab when I go to load. I am on v3.1.0-beta.6
 [2026-01-01T21:38:27.237Z] [INFO] Images synced: 547,596 in 353.5s
[2026-01-01T21:38:27.242Z] [INFO] Rebuilding inherited image counts...
[2026-01-01T21:38:27.242Z] [INFO] Rebuilding inherited image counts for all entities...
[2026-01-01T21:39:33.662Z] [ERROR] Failed to rebuild inherited image counts {"error":"\nInvalid `prisma.stashImage.findMany()` invocation:\n\n\nFailed to convert rust `String` into napi `string`"}
[2026-01-01T21:39:33.668Z] [ERROR] Full sync failed {"error":"\nInvalid `prisma.stashImage.findMany()` invocation:\n\n\nFailed to convert rust `String` into napi `string`"}
[2026-01-01T21:39:33.668Z] [ERROR] Background full sync failed {"error":"\nInvalid `prisma.stashImage.findMany()` invocation:\n\n\nFailed to convert rust `String` into napi `string`"}

--- a/server/controllers/library/galleries.ts
+++ b/server/controllers/library/galleries.ts
@@ -605,6 +605,11 @@ export const getGalleryImages = async (
     const { galleryId } = req.params;
     const userId = req.user?.id;
 
+    // Pagination parameters (optional - defaults to loading all for backwards compat)
+    const page = parseInt(req.query.page as string) || 1;
+    const perPage = parseInt(req.query.per_page as string) || 0; // 0 = no limit (all)
+    const skip = perPage > 0 ? (page - 1) * perPage : 0;
+
     if (!userId) {
       return res.status(401).json({ error: "Unauthorized" });
     }
@@ -612,20 +617,30 @@ export const getGalleryImages = async (
     // Fetch the gallery data for inheritance context
     const gallery = await stashEntityService.getGallery(galleryId);
 
-    // Query images from local database filtered by gallery
-    const dbImages = await prisma.stashImage.findMany({
-      where: {
-        deletedAt: null,
-        galleries: {
-          some: { galleryId },
-        },
+    // Build query options
+    const whereClause = {
+      deletedAt: null,
+      galleries: {
+        some: { galleryId },
       },
+    };
+
+    // Get total count for pagination metadata
+    const totalCount = perPage > 0
+      ? await prisma.stashImage.count({ where: whereClause })
+      : 0;
+
+    // Query images from local database filtered by gallery
+    // PERFORMANCE: Uses pagination when per_page is specified to avoid loading huge galleries
+    const dbImages = await prisma.stashImage.findMany({
+      where: whereClause,
       include: {
         performers: { include: { performer: true } },
         tags: { include: { tag: true } },
         studio: true,
       },
       orderBy: { filePath: "asc" },
+      ...(perPage > 0 && { skip, take: perPage }),
     });
 
     // Build gallery context for inheritance (used by client-side getEffectiveImageMetadata)
@@ -721,10 +736,32 @@ export const getGalleryImages = async (
       userId
     );
 
-    res.json({
+    // Build response with optional pagination metadata
+    const response: {
+      images: typeof mergedImages;
+      count: number;
+      pagination?: {
+        page: number;
+        per_page: number;
+        total: number;
+        total_pages: number;
+      };
+    } = {
       images: mergedImages,
       count: mergedImages.length,
-    });
+    };
+
+    // Include pagination metadata when per_page is specified
+    if (perPage > 0) {
+      response.pagination = {
+        page,
+        per_page: perPage,
+        total: totalCount,
+        total_pages: Math.ceil(totalCount / perPage),
+      };
+    }
+
+    res.json(response);
   } catch (error) {
     logger.error("Error fetching gallery images", {
       galleryId: req.params.galleryId,

--- a/server/controllers/library/scenes.ts
+++ b/server/controllers/library/scenes.ts
@@ -52,12 +52,21 @@ class SeededRandom {
 
 /**
  * Merge user-specific data into scenes
+ *
+ * PERFORMANCE: When fetching data for a small number of scenes (< 100),
+ * we filter by sceneId to avoid loading entire watch history tables.
+ * For larger sets, we load all data and use Map lookups for efficiency.
  */
 export async function mergeScenesWithUserData(
   scenes: NormalizedScene[],
   userId: number
 ): Promise<NormalizedScene[]> {
+  // Extract scene IDs for targeted queries when dealing with small sets
+  const sceneIds = scenes.map((s) => s.id);
+  const useTargetedQuery = sceneIds.length < 100;
+
   // Fetch user data in parallel
+  // For small scene sets, filter by sceneId to avoid loading full tables
   const [
     watchHistory,
     sceneRatings,
@@ -65,8 +74,18 @@ export async function mergeScenesWithUserData(
     studioRatings,
     tagRatings,
   ] = await Promise.all([
-    prisma.watchHistory.findMany({ where: { userId } }),
-    prisma.sceneRating.findMany({ where: { userId } }),
+    prisma.watchHistory.findMany({
+      where: useTargetedQuery
+        ? { userId, sceneId: { in: sceneIds } }
+        : { userId },
+    }),
+    prisma.sceneRating.findMany({
+      where: useTargetedQuery
+        ? { userId, sceneId: { in: sceneIds } }
+        : { userId },
+    }),
+    // Performer/studio/tag ratings are kept as full loads since they're
+    // used for nested entity favorites across all scenes
     prisma.performerRating.findMany({ where: { userId } }),
     prisma.studioRating.findMany({ where: { userId } }),
     prisma.tagRating.findMany({ where: { userId } }),

--- a/server/services/EntityImageCountService.ts
+++ b/server/services/EntityImageCountService.ts
@@ -1,6 +1,5 @@
 import prisma from "../prisma/singleton.js";
 import { logger } from "../utils/logger.js";
-import { stashEntityService } from "./StashEntityService.js";
 
 /**
  * EntityImageCountService
@@ -16,39 +15,28 @@ import { stashEntityService } from "./StashEntityService.js";
  *
  * This service recalculates these counts after sync and stores them in the database
  * for fast retrieval on detail pages.
+ *
+ * PERFORMANCE: Uses SQL aggregation instead of loading all images into memory.
+ * This scales to millions of images without memory issues.
  */
-
-interface ImageWithRelations {
-  id: string;
-  studioId?: string | null;
-  performers?: Array<{ id: string }>;
-  tags?: Array<{ id: string }>;
-  galleries?: Array<{
-    id: string;
-    studioId?: string | null;
-    performers?: Array<{ id: string }>;
-    tags?: Array<{ id: string }>;
-  }>;
-}
 
 class EntityImageCountService {
   /**
    * Rebuild inherited image counts for all entity types.
    * Called after sync completes to ensure counts reflect gallery inheritance.
+   *
+   * Uses SQL UNION queries to count images efficiently without loading them into memory.
    */
   async rebuildAllImageCounts(): Promise<void> {
     const startTime = Date.now();
     logger.info("Rebuilding inherited image counts for all entities...");
 
     try {
-      // Load all images with their relationships once
-      const allImages = await stashEntityService.getAllImages();
-
-      // Rebuild counts for each entity type in parallel
+      // Rebuild counts for each entity type in parallel using SQL aggregation
       await Promise.all([
-        this.rebuildPerformerImageCounts(allImages),
-        this.rebuildStudioImageCounts(allImages),
-        this.rebuildTagImageCounts(allImages),
+        this.rebuildPerformerImageCountsSQL(),
+        this.rebuildStudioImageCountsSQL(),
+        this.rebuildTagImageCountsSQL(),
       ]);
 
       const duration = Date.now() - startTime;
@@ -62,229 +50,141 @@ class EntityImageCountService {
   }
 
   /**
-   * Rebuild image counts for all performers.
+   * Rebuild image counts for all performers using SQL aggregation.
    * Counts images where performer is directly tagged OR tagged on a parent gallery.
+   *
+   * Uses UNION to combine:
+   * 1. Direct: ImagePerformer joins
+   * 2. Inherited: ImageGallery -> GalleryPerformer joins
    */
-  async rebuildPerformerImageCounts(allImages?: ImageWithRelations[]): Promise<void> {
+  private async rebuildPerformerImageCountsSQL(): Promise<void> {
     const startTime = Date.now();
 
-    // Load images if not provided
-    const images = allImages ?? await stashEntityService.getAllImages();
+    // Single SQL query that:
+    // 1. Finds all distinct (imageId, performerId) pairs from direct + inherited relations
+    // 2. Groups by performerId to get counts
+    // 3. Updates all performers in one batch
+    await prisma.$executeRaw`
+      UPDATE StashPerformer
+      SET imageCount = COALESCE((
+        SELECT COUNT(DISTINCT imageId) FROM (
+          -- Direct: image has performer directly
+          SELECT ip.imageId, ip.performerId
+          FROM ImagePerformer ip
+          JOIN StashImage si ON ip.imageId = si.id AND si.deletedAt IS NULL
 
-    // Get all performer IDs
-    const performers = await prisma.stashPerformer.findMany({
-      where: { deletedAt: null },
-      select: { id: true },
-    });
+          UNION
 
-    if (performers.length === 0) {
-      logger.debug("No performers to update image counts for");
-      return;
-    }
-
-    // Calculate counts for each performer
-    const counts = new Map<string, number>();
-
-    for (const performer of performers) {
-      const count = this.countImagesForPerformer(images, performer.id);
-      counts.set(performer.id, count);
-    }
-
-    // Batch update all performers
-    await this.batchUpdatePerformerImageCounts(counts);
+          -- Inherited: image is in gallery that has performer
+          SELECT ig.imageId, gp.performerId
+          FROM ImageGallery ig
+          JOIN StashImage si ON ig.imageId = si.id AND si.deletedAt IS NULL
+          JOIN StashGallery sg ON ig.galleryId = sg.id AND sg.deletedAt IS NULL
+          JOIN GalleryPerformer gp ON ig.galleryId = gp.galleryId
+        ) AS combined
+        WHERE combined.performerId = StashPerformer.id
+      ), 0)
+      WHERE StashPerformer.deletedAt IS NULL
+    `;
 
     const duration = Date.now() - startTime;
-    logger.debug(`Performer image counts rebuilt in ${duration}ms`, {
-      performerCount: performers.length,
-    });
+    logger.debug(`Performer image counts rebuilt via SQL in ${duration}ms`);
   }
 
   /**
-   * Rebuild image counts for all studios.
+   * Rebuild image counts for all studios using SQL aggregation.
    * Counts images where studio is directly set OR set on a parent gallery.
    */
-  async rebuildStudioImageCounts(allImages?: ImageWithRelations[]): Promise<void> {
+  private async rebuildStudioImageCountsSQL(): Promise<void> {
     const startTime = Date.now();
 
-    const images = allImages ?? await stashEntityService.getAllImages();
+    await prisma.$executeRaw`
+      UPDATE StashStudio
+      SET imageCount = COALESCE((
+        SELECT COUNT(DISTINCT imageId) FROM (
+          -- Direct: image has studio directly
+          SELECT si.id AS imageId, si.studioId
+          FROM StashImage si
+          WHERE si.deletedAt IS NULL AND si.studioId IS NOT NULL
 
-    const studios = await prisma.stashStudio.findMany({
-      where: { deletedAt: null },
-      select: { id: true },
-    });
+          UNION
 
-    if (studios.length === 0) {
-      logger.debug("No studios to update image counts for");
-      return;
-    }
-
-    const counts = new Map<string, number>();
-
-    for (const studio of studios) {
-      const count = this.countImagesForStudio(images, studio.id);
-      counts.set(studio.id, count);
-    }
-
-    await this.batchUpdateStudioImageCounts(counts);
+          -- Inherited: image is in gallery that has studio
+          SELECT ig.imageId, sg.studioId
+          FROM ImageGallery ig
+          JOIN StashImage si ON ig.imageId = si.id AND si.deletedAt IS NULL
+          JOIN StashGallery sg ON ig.galleryId = sg.id AND sg.deletedAt IS NULL AND sg.studioId IS NOT NULL
+        ) AS combined
+        WHERE combined.studioId = StashStudio.id
+      ), 0)
+      WHERE StashStudio.deletedAt IS NULL
+    `;
 
     const duration = Date.now() - startTime;
-    logger.debug(`Studio image counts rebuilt in ${duration}ms`, {
-      studioCount: studios.length,
-    });
+    logger.debug(`Studio image counts rebuilt via SQL in ${duration}ms`);
   }
 
   /**
-   * Rebuild image counts for all tags.
+   * Rebuild image counts for all tags using SQL aggregation.
    * Counts images where tag is directly set OR set on a parent gallery.
    */
-  async rebuildTagImageCounts(allImages?: ImageWithRelations[]): Promise<void> {
+  private async rebuildTagImageCountsSQL(): Promise<void> {
     const startTime = Date.now();
 
-    const images = allImages ?? await stashEntityService.getAllImages();
+    await prisma.$executeRaw`
+      UPDATE StashTag
+      SET imageCount = COALESCE((
+        SELECT COUNT(DISTINCT imageId) FROM (
+          -- Direct: image has tag directly
+          SELECT it.imageId, it.tagId
+          FROM ImageTag it
+          JOIN StashImage si ON it.imageId = si.id AND si.deletedAt IS NULL
 
-    const tags = await prisma.stashTag.findMany({
-      where: { deletedAt: null },
-      select: { id: true },
-    });
+          UNION
 
-    if (tags.length === 0) {
-      logger.debug("No tags to update image counts for");
-      return;
-    }
-
-    const counts = new Map<string, number>();
-
-    for (const tag of tags) {
-      const count = this.countImagesForTag(images, tag.id);
-      counts.set(tag.id, count);
-    }
-
-    await this.batchUpdateTagImageCounts(counts);
+          -- Inherited: image is in gallery that has tag
+          SELECT ig.imageId, gt.tagId
+          FROM ImageGallery ig
+          JOIN StashImage si ON ig.imageId = si.id AND si.deletedAt IS NULL
+          JOIN StashGallery sg ON ig.galleryId = sg.id AND sg.deletedAt IS NULL
+          JOIN GalleryTag gt ON ig.galleryId = gt.galleryId
+        ) AS combined
+        WHERE combined.tagId = StashTag.id
+      ), 0)
+      WHERE StashTag.deletedAt IS NULL
+    `;
 
     const duration = Date.now() - startTime;
-    logger.debug(`Tag image counts rebuilt in ${duration}ms`, {
-      tagCount: tags.length,
-    });
+    logger.debug(`Tag image counts rebuilt via SQL in ${duration}ms`);
+  }
+
+  // ============================================================================
+  // Legacy methods kept for backwards compatibility / individual entity updates
+  // These use the old approach but are only called for single-entity updates
+  // ============================================================================
+
+  /**
+   * Rebuild image counts for all performers (legacy method).
+   * @deprecated Use rebuildPerformerImageCountsSQL for bulk operations
+   */
+  async rebuildPerformerImageCounts(): Promise<void> {
+    return this.rebuildPerformerImageCountsSQL();
   }
 
   /**
-   * Count images for a performer using gallery inheritance.
+   * Rebuild image counts for all studios (legacy method).
+   * @deprecated Use rebuildStudioImageCountsSQL for bulk operations
    */
-  private countImagesForPerformer(images: ImageWithRelations[], performerId: string): number {
-    return images.filter((img) => {
-      // Check direct performers
-      if (img.performers?.some((p) => String(p.id) === String(performerId))) {
-        return true;
-      }
-      // Check gallery performers (inheritance)
-      if (img.galleries?.some((g) =>
-        g.performers?.some((p) => String(p.id) === String(performerId))
-      )) {
-        return true;
-      }
-      return false;
-    }).length;
+  async rebuildStudioImageCounts(): Promise<void> {
+    return this.rebuildStudioImageCountsSQL();
   }
 
   /**
-   * Count images for a studio using gallery inheritance.
+   * Rebuild image counts for all tags (legacy method).
+   * @deprecated Use rebuildTagImageCountsSQL for bulk operations
    */
-  private countImagesForStudio(images: ImageWithRelations[], studioId: string): number {
-    return images.filter((img) => {
-      // Check direct studio
-      if (String(img.studioId) === String(studioId)) {
-        return true;
-      }
-      // Check gallery studio (inheritance)
-      if (img.galleries?.some((g) => String(g.studioId) === String(studioId))) {
-        return true;
-      }
-      return false;
-    }).length;
-  }
-
-  /**
-   * Count images for a tag using gallery inheritance.
-   */
-  private countImagesForTag(images: ImageWithRelations[], tagId: string): number {
-    return images.filter((img) => {
-      // Check direct tags
-      if (img.tags?.some((t) => String(t.id) === String(tagId))) {
-        return true;
-      }
-      // Check gallery tags (inheritance)
-      if (img.galleries?.some((g) =>
-        g.tags?.some((t) => String(t.id) === String(tagId))
-      )) {
-        return true;
-      }
-      return false;
-    }).length;
-  }
-
-  /**
-   * Batch update performer image counts in the database.
-   * Uses raw SQL for performance with large datasets.
-   */
-  private async batchUpdatePerformerImageCounts(counts: Map<string, number>): Promise<void> {
-    if (counts.size === 0) return;
-
-    // Build CASE statement for batch update
-    const cases = Array.from(counts.entries())
-      .map(([id, count]) => `WHEN '${id}' THEN ${count}`)
-      .join(" ");
-
-    const ids = Array.from(counts.keys())
-      .map((id) => `'${id}'`)
-      .join(", ");
-
-    await prisma.$executeRawUnsafe(`
-      UPDATE StashPerformer
-      SET imageCount = CASE id ${cases} ELSE imageCount END
-      WHERE id IN (${ids})
-    `);
-  }
-
-  /**
-   * Batch update studio image counts in the database.
-   */
-  private async batchUpdateStudioImageCounts(counts: Map<string, number>): Promise<void> {
-    if (counts.size === 0) return;
-
-    const cases = Array.from(counts.entries())
-      .map(([id, count]) => `WHEN '${id}' THEN ${count}`)
-      .join(" ");
-
-    const ids = Array.from(counts.keys())
-      .map((id) => `'${id}'`)
-      .join(", ");
-
-    await prisma.$executeRawUnsafe(`
-      UPDATE StashStudio
-      SET imageCount = CASE id ${cases} ELSE imageCount END
-      WHERE id IN (${ids})
-    `);
-  }
-
-  /**
-   * Batch update tag image counts in the database.
-   */
-  private async batchUpdateTagImageCounts(counts: Map<string, number>): Promise<void> {
-    if (counts.size === 0) return;
-
-    const cases = Array.from(counts.entries())
-      .map(([id, count]) => `WHEN '${id}' THEN ${count}`)
-      .join(" ");
-
-    const ids = Array.from(counts.keys())
-      .map((id) => `'${id}'`)
-      .join(", ");
-
-    await prisma.$executeRawUnsafe(`
-      UPDATE StashTag
-      SET imageCount = CASE id ${cases} ELSE imageCount END
-      WHERE id IN (${ids})
-    `);
+  async rebuildTagImageCounts(): Promise<void> {
+    return this.rebuildTagImageCountsSQL();
   }
 }
 

--- a/server/services/UserRestrictionService.ts
+++ b/server/services/UserRestrictionService.ts
@@ -56,6 +56,9 @@ class UserRestrictionService {
    * 5. Scenes with hidden groups (CASCADE)
    * 6. Scenes with hidden galleries (CASCADE)
    * 7. Content restrictions (INCLUDE/EXCLUDE rules) - unless skipContentRestrictions
+   *
+   * PERFORMANCE: Caches allSceneIds query to avoid redundant full table scans.
+   * For INCLUDE restrictions, we only load the full scene list once.
    */
   async getExcludedSceneIds(
     userId: number,
@@ -139,12 +142,27 @@ class UserRestrictionService {
     if (!skipContentRestrictions) {
       const restrictions = await this.getRestrictionsForUser(userId);
 
+      // PERFORMANCE: Check if we need allSceneIds (only for INCLUDE mode or restrictEmpty)
+      // Cache it once instead of querying multiple times
+      let allSceneIdsCache: Set<string> | null = null;
+      const needsAllSceneIds = restrictions.some(
+        (r) => r.mode === "INCLUDE" || r.restrictEmpty
+      );
+
+      if (needsAllSceneIds) {
+        const allScenes = await prisma.stashScene.findMany({
+          where: { deletedAt: null },
+          select: { id: true },
+        });
+        allSceneIdsCache = new Set(allScenes.map((s) => s.id));
+      }
+
       for (const restriction of restrictions) {
         const entityIds = JSON.parse(restriction.entityIds) as string[];
-        if (entityIds.length === 0) continue;
+        if (entityIds.length === 0 && !restriction.restrictEmpty) continue;
 
         if (restriction.entityType === "groups") {
-          if (restriction.mode === "EXCLUDE") {
+          if (restriction.mode === "EXCLUDE" && entityIds.length > 0) {
             // Exclude scenes in these groups
             const scenesInExcludedGroups = await prisma.sceneGroup.findMany({
               where: { groupId: { in: entityIds } },
@@ -153,28 +171,38 @@ class UserRestrictionService {
             for (const sg of scenesInExcludedGroups) {
               excludedIds.add(sg.sceneId);
             }
-          } else if (restriction.mode === "INCLUDE") {
+          } else if (restriction.mode === "INCLUDE" && entityIds.length > 0) {
             // Only include scenes in these groups - exclude all others
-            // Get all scene IDs that ARE in the included groups
             const scenesInIncludedGroups = await prisma.sceneGroup.findMany({
               where: { groupId: { in: entityIds } },
               select: { sceneId: true },
             });
             const includedSceneIds = new Set(scenesInIncludedGroups.map(sg => sg.sceneId));
 
-            // Get all scene IDs and exclude those NOT in the included groups
-            const allSceneIds = await prisma.stashScene.findMany({
-              where: { deletedAt: null },
-              select: { id: true },
+            // Use cached allSceneIds
+            for (const sceneId of allSceneIdsCache!) {
+              if (!includedSceneIds.has(sceneId)) {
+                excludedIds.add(sceneId);
+              }
+            }
+          }
+
+          // Handle restrictEmpty for groups
+          if (restriction.restrictEmpty) {
+            const scenesWithGroups = await prisma.sceneGroup.findMany({
+              select: { sceneId: true },
+              distinct: ['sceneId'],
             });
-            for (const s of allSceneIds) {
-              if (!includedSceneIds.has(s.id)) {
-                excludedIds.add(s.id);
+            const scenesWithGroupsSet = new Set(scenesWithGroups.map(sg => sg.sceneId));
+
+            for (const sceneId of allSceneIdsCache!) {
+              if (!scenesWithGroupsSet.has(sceneId)) {
+                excludedIds.add(sceneId);
               }
             }
           }
         } else if (restriction.entityType === "tags") {
-          if (restriction.mode === "EXCLUDE") {
+          if (restriction.mode === "EXCLUDE" && entityIds.length > 0) {
             // Exclude scenes with these tags
             const scenesWithExcludedTags = await prisma.sceneTag.findMany({
               where: { tagId: { in: entityIds } },
@@ -183,7 +211,7 @@ class UserRestrictionService {
             for (const st of scenesWithExcludedTags) {
               excludedIds.add(st.sceneId);
             }
-          } else if (restriction.mode === "INCLUDE") {
+          } else if (restriction.mode === "INCLUDE" && entityIds.length > 0) {
             // Only include scenes with these tags - exclude all others
             const scenesWithIncludedTags = await prisma.sceneTag.findMany({
               where: { tagId: { in: entityIds } },
@@ -191,18 +219,29 @@ class UserRestrictionService {
             });
             const includedSceneIds = new Set(scenesWithIncludedTags.map(st => st.sceneId));
 
-            const allSceneIds = await prisma.stashScene.findMany({
-              where: { deletedAt: null },
-              select: { id: true },
+            for (const sceneId of allSceneIdsCache!) {
+              if (!includedSceneIds.has(sceneId)) {
+                excludedIds.add(sceneId);
+              }
+            }
+          }
+
+          // Handle restrictEmpty for tags
+          if (restriction.restrictEmpty) {
+            const scenesWithTags = await prisma.sceneTag.findMany({
+              select: { sceneId: true },
+              distinct: ['sceneId'],
             });
-            for (const s of allSceneIds) {
-              if (!includedSceneIds.has(s.id)) {
-                excludedIds.add(s.id);
+            const scenesWithTagsSet = new Set(scenesWithTags.map(st => st.sceneId));
+
+            for (const sceneId of allSceneIdsCache!) {
+              if (!scenesWithTagsSet.has(sceneId)) {
+                excludedIds.add(sceneId);
               }
             }
           }
         } else if (restriction.entityType === "studios") {
-          if (restriction.mode === "EXCLUDE") {
+          if (restriction.mode === "EXCLUDE" && entityIds.length > 0) {
             // Exclude scenes from these studios
             const scenesFromExcludedStudios = await prisma.stashScene.findMany({
               where: {
@@ -214,7 +253,7 @@ class UserRestrictionService {
             for (const s of scenesFromExcludedStudios) {
               excludedIds.add(s.id);
             }
-          } else if (restriction.mode === "INCLUDE") {
+          } else if (restriction.mode === "INCLUDE" && entityIds.length > 0) {
             // Only include scenes from these studios - exclude all others
             const scenesFromIncludedStudios = await prisma.stashScene.findMany({
               where: {
@@ -225,84 +264,15 @@ class UserRestrictionService {
             });
             const includedSceneIds = new Set(scenesFromIncludedStudios.map(s => s.id));
 
-            const allSceneIds = await prisma.stashScene.findMany({
-              where: { deletedAt: null },
-              select: { id: true },
-            });
-            for (const s of allSceneIds) {
-              if (!includedSceneIds.has(s.id)) {
-                excludedIds.add(s.id);
+            for (const sceneId of allSceneIdsCache!) {
+              if (!includedSceneIds.has(sceneId)) {
+                excludedIds.add(sceneId);
               }
             }
           }
-        } else if (restriction.entityType === "galleries") {
-          if (restriction.mode === "EXCLUDE") {
-            // Exclude scenes linked to these galleries
-            const scenesWithExcludedGalleries = await prisma.sceneGallery.findMany({
-              where: { galleryId: { in: entityIds } },
-              select: { sceneId: true },
-            });
-            for (const sg of scenesWithExcludedGalleries) {
-              excludedIds.add(sg.sceneId);
-            }
-          } else if (restriction.mode === "INCLUDE") {
-            // Only include scenes linked to these galleries - exclude all others
-            const scenesWithIncludedGalleries = await prisma.sceneGallery.findMany({
-              where: { galleryId: { in: entityIds } },
-              select: { sceneId: true },
-            });
-            const includedSceneIds = new Set(scenesWithIncludedGalleries.map(sg => sg.sceneId));
 
-            const allSceneIds = await prisma.stashScene.findMany({
-              where: { deletedAt: null },
-              select: { id: true },
-            });
-            for (const s of allSceneIds) {
-              if (!includedSceneIds.has(s.id)) {
-                excludedIds.add(s.id);
-              }
-            }
-          }
-        }
-
-        // Handle restrictEmpty (exclude scenes with no entities of this type)
-        if (restriction.restrictEmpty) {
-          if (restriction.entityType === "groups") {
-            // Find scenes with no groups
-            const scenesWithGroups = await prisma.sceneGroup.findMany({
-              select: { sceneId: true },
-              distinct: ['sceneId'],
-            });
-            const scenesWithGroupsSet = new Set(scenesWithGroups.map(sg => sg.sceneId));
-
-            const allSceneIds = await prisma.stashScene.findMany({
-              where: { deletedAt: null },
-              select: { id: true },
-            });
-            for (const s of allSceneIds) {
-              if (!scenesWithGroupsSet.has(s.id)) {
-                excludedIds.add(s.id);
-              }
-            }
-          } else if (restriction.entityType === "tags") {
-            // Find scenes with no tags
-            const scenesWithTags = await prisma.sceneTag.findMany({
-              select: { sceneId: true },
-              distinct: ['sceneId'],
-            });
-            const scenesWithTagsSet = new Set(scenesWithTags.map(st => st.sceneId));
-
-            const allSceneIds = await prisma.stashScene.findMany({
-              where: { deletedAt: null },
-              select: { id: true },
-            });
-            for (const s of allSceneIds) {
-              if (!scenesWithTagsSet.has(s.id)) {
-                excludedIds.add(s.id);
-              }
-            }
-          } else if (restriction.entityType === "studios") {
-            // Find scenes with no studio
+          // Handle restrictEmpty for studios
+          if (restriction.restrictEmpty) {
             const scenesWithoutStudio = await prisma.stashScene.findMany({
               where: {
                 OR: [
@@ -316,21 +286,43 @@ class UserRestrictionService {
             for (const s of scenesWithoutStudio) {
               excludedIds.add(s.id);
             }
-          } else if (restriction.entityType === "galleries") {
-            // Find scenes with no galleries
+          }
+        } else if (restriction.entityType === "galleries") {
+          if (restriction.mode === "EXCLUDE" && entityIds.length > 0) {
+            // Exclude scenes linked to these galleries
+            const scenesWithExcludedGalleries = await prisma.sceneGallery.findMany({
+              where: { galleryId: { in: entityIds } },
+              select: { sceneId: true },
+            });
+            for (const sg of scenesWithExcludedGalleries) {
+              excludedIds.add(sg.sceneId);
+            }
+          } else if (restriction.mode === "INCLUDE" && entityIds.length > 0) {
+            // Only include scenes linked to these galleries - exclude all others
+            const scenesWithIncludedGalleries = await prisma.sceneGallery.findMany({
+              where: { galleryId: { in: entityIds } },
+              select: { sceneId: true },
+            });
+            const includedSceneIds = new Set(scenesWithIncludedGalleries.map(sg => sg.sceneId));
+
+            for (const sceneId of allSceneIdsCache!) {
+              if (!includedSceneIds.has(sceneId)) {
+                excludedIds.add(sceneId);
+              }
+            }
+          }
+
+          // Handle restrictEmpty for galleries
+          if (restriction.restrictEmpty) {
             const scenesWithGalleries = await prisma.sceneGallery.findMany({
               select: { sceneId: true },
               distinct: ['sceneId'],
             });
             const scenesWithGalleriesSet = new Set(scenesWithGalleries.map(sg => sg.sceneId));
 
-            const allSceneIds = await prisma.stashScene.findMany({
-              where: { deletedAt: null },
-              select: { id: true },
-            });
-            for (const s of allSceneIds) {
-              if (!scenesWithGalleriesSet.has(s.id)) {
-                excludedIds.add(s.id);
+            for (const sceneId of allSceneIdsCache!) {
+              if (!scenesWithGalleriesSet.has(sceneId)) {
+                excludedIds.add(sceneId);
               }
             }
           }

--- a/server/services/__tests__/EntityImageCountService.test.ts
+++ b/server/services/__tests__/EntityImageCountService.test.ts
@@ -2,7 +2,11 @@
  * Unit Tests for EntityImageCountService
  *
  * Tests the inherited image count calculation for performers, studios, and tags.
- * Verifies that gallery-inherited images are properly counted.
+ * The service now uses SQL aggregation queries instead of in-memory filtering.
+ *
+ * These tests verify that the SQL queries are executed correctly.
+ * The actual count logic is handled by the database, so we're primarily
+ * testing that the service calls the correct Prisma methods.
  */
 import { describe, it, expect, beforeEach, vi, type Mock } from "vitest";
 
@@ -23,28 +27,11 @@ vi.mock("../StashInstanceManager.js", () => ({
 // Mock Prisma
 vi.mock("../../prisma/singleton.js", () => ({
   default: {
-    stashPerformer: {
-      findMany: vi.fn(),
-    },
-    stashStudio: {
-      findMany: vi.fn(),
-    },
-    stashTag: {
-      findMany: vi.fn(),
-    },
-    $executeRawUnsafe: vi.fn(),
-  },
-}));
-
-// Mock StashEntityService
-vi.mock("../StashEntityService.js", () => ({
-  stashEntityService: {
-    getAllImages: vi.fn(),
+    $executeRaw: vi.fn().mockResolvedValue(0),
   },
 }));
 
 import prisma from "../../prisma/singleton.js";
-import { stashEntityService } from "../StashEntityService.js";
 import { entityImageCountService } from "../EntityImageCountService.js";
 
 const getMock = (fn: unknown): Mock => fn as Mock;
@@ -54,287 +41,37 @@ describe("EntityImageCountService", () => {
     vi.clearAllMocks();
   });
 
-  describe("rebuildPerformerImageCounts", () => {
-    it("counts direct performer-image associations", async () => {
-      // Setup: Performer "p1" is directly tagged on images
-      getMock(prisma.stashPerformer.findMany).mockResolvedValue([
-        { id: "p1" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          performers: [{ id: "p1" }],
-          galleries: [],
-        },
-        {
-          id: "img2",
-          performers: [{ id: "p1" }],
-          galleries: [],
-        },
-        {
-          id: "img3",
-          performers: [{ id: "p2" }], // Different performer
-          galleries: [],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
+  describe("rebuildPerformerImageCountsSQL", () => {
+    it("executes SQL to update performer image counts", async () => {
       await entityImageCountService.rebuildPerformerImageCounts();
 
-      // Should update with count of 2 (img1 and img2)
-      expect(prisma.$executeRawUnsafe).toHaveBeenCalledWith(
-        expect.stringContaining("WHEN 'p1' THEN 2")
-      );
-    });
-
-    it("counts gallery-inherited performer associations", async () => {
-      // Setup: Performer "p1" is tagged on a gallery, not directly on images
-      getMock(prisma.stashPerformer.findMany).mockResolvedValue([
-        { id: "p1" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          performers: [], // No direct performers
-          galleries: [
-            {
-              id: "g1",
-              performers: [{ id: "p1" }], // Performer on gallery
-            },
-          ],
-        },
-        {
-          id: "img2",
-          performers: [],
-          galleries: [
-            {
-              id: "g1",
-              performers: [{ id: "p1" }],
-            },
-          ],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
-      await entityImageCountService.rebuildPerformerImageCounts();
-
-      // Should count both images via gallery inheritance
-      expect(prisma.$executeRawUnsafe).toHaveBeenCalledWith(
-        expect.stringContaining("WHEN 'p1' THEN 2")
-      );
-    });
-
-    it("counts both direct and inherited associations without duplicates", async () => {
-      // Setup: Performer on both image directly AND on the gallery
-      getMock(prisma.stashPerformer.findMany).mockResolvedValue([
-        { id: "p1" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          performers: [{ id: "p1" }], // Direct
-          galleries: [
-            {
-              id: "g1",
-              performers: [{ id: "p1" }], // Also on gallery
-            },
-          ],
-        },
-        {
-          id: "img2",
-          performers: [], // Only via gallery
-          galleries: [
-            {
-              id: "g1",
-              performers: [{ id: "p1" }],
-            },
-          ],
-        },
-        {
-          id: "img3",
-          performers: [{ id: "p1" }], // Direct only
-          galleries: [],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
-      await entityImageCountService.rebuildPerformerImageCounts();
-
-      // Should count 3 images (no duplicates)
-      expect(prisma.$executeRawUnsafe).toHaveBeenCalledWith(
-        expect.stringContaining("WHEN 'p1' THEN 3")
-      );
-    });
-
-    it("handles performers with zero images", async () => {
-      getMock(prisma.stashPerformer.findMany).mockResolvedValue([
-        { id: "p1" },
-        { id: "p2" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          performers: [{ id: "p1" }],
-          galleries: [],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
-      await entityImageCountService.rebuildPerformerImageCounts();
-
-      // p1 should have 1, p2 should have 0
-      const call = getMock(prisma.$executeRawUnsafe).mock.calls[0][0];
-      expect(call).toContain("WHEN 'p1' THEN 1");
-      expect(call).toContain("WHEN 'p2' THEN 0");
+      // Verify $executeRaw was called (the SQL aggregation query)
+      expect(prisma.$executeRaw).toHaveBeenCalledTimes(1);
     });
   });
 
-  describe("rebuildStudioImageCounts", () => {
-    it("counts direct studio-image associations", async () => {
-      getMock(prisma.stashStudio.findMany).mockResolvedValue([
-        { id: "s1" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          studioId: "s1",
-          galleries: [],
-        },
-        {
-          id: "img2",
-          studioId: "s1",
-          galleries: [],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
+  describe("rebuildStudioImageCountsSQL", () => {
+    it("executes SQL to update studio image counts", async () => {
       await entityImageCountService.rebuildStudioImageCounts();
 
-      expect(prisma.$executeRawUnsafe).toHaveBeenCalledWith(
-        expect.stringContaining("WHEN 's1' THEN 2")
-      );
-    });
-
-    it("counts gallery-inherited studio associations", async () => {
-      getMock(prisma.stashStudio.findMany).mockResolvedValue([
-        { id: "s1" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          studioId: null, // No direct studio
-          galleries: [
-            {
-              id: "g1",
-              studioId: "s1", // Studio on gallery
-            },
-          ],
-        },
-        {
-          id: "img2",
-          studioId: null,
-          galleries: [
-            {
-              id: "g1",
-              studioId: "s1",
-            },
-          ],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
-      await entityImageCountService.rebuildStudioImageCounts();
-
-      expect(prisma.$executeRawUnsafe).toHaveBeenCalledWith(
-        expect.stringContaining("WHEN 's1' THEN 2")
-      );
+      expect(prisma.$executeRaw).toHaveBeenCalledTimes(1);
     });
   });
 
-  describe("rebuildTagImageCounts", () => {
-    it("counts direct tag-image associations", async () => {
-      getMock(prisma.stashTag.findMany).mockResolvedValue([
-        { id: "t1" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          tags: [{ id: "t1" }],
-          galleries: [],
-        },
-        {
-          id: "img2",
-          tags: [{ id: "t1" }, { id: "t2" }],
-          galleries: [],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
+  describe("rebuildTagImageCountsSQL", () => {
+    it("executes SQL to update tag image counts", async () => {
       await entityImageCountService.rebuildTagImageCounts();
 
-      expect(prisma.$executeRawUnsafe).toHaveBeenCalledWith(
-        expect.stringContaining("WHEN 't1' THEN 2")
-      );
-    });
-
-    it("counts gallery-inherited tag associations", async () => {
-      getMock(prisma.stashTag.findMany).mockResolvedValue([
-        { id: "t1" },
-      ]);
-
-      getMock(stashEntityService.getAllImages).mockResolvedValue([
-        {
-          id: "img1",
-          tags: [],
-          galleries: [
-            {
-              id: "g1",
-              tags: [{ id: "t1" }],
-            },
-          ],
-        },
-      ]);
-
-      getMock(prisma.$executeRawUnsafe).mockResolvedValue(undefined);
-
-      await entityImageCountService.rebuildTagImageCounts();
-
-      expect(prisma.$executeRawUnsafe).toHaveBeenCalledWith(
-        expect.stringContaining("WHEN 't1' THEN 1")
-      );
+      expect(prisma.$executeRaw).toHaveBeenCalledTimes(1);
     });
   });
 
   describe("rebuildAllImageCounts", () => {
-    it("rebuilds counts for all entity types", async () => {
-      getMock(prisma.stashPerformer.findMany).mockResolvedValue([]);
-      getMock(prisma.stashStudio.findMany).mockResolvedValue([]);
-      getMock(prisma.stashTag.findMany).mockResolvedValue([]);
-      getMock(stashEntityService.getAllImages).mockResolvedValue([]);
-
+    it("rebuilds counts for all entity types in parallel", async () => {
       await entityImageCountService.rebuildAllImageCounts();
 
-      // getAllImages should be called once (shared across all entity types)
-      expect(stashEntityService.getAllImages).toHaveBeenCalledTimes(1);
-
-      // All entity types should be queried
-      expect(prisma.stashPerformer.findMany).toHaveBeenCalled();
-      expect(prisma.stashStudio.findMany).toHaveBeenCalled();
-      expect(prisma.stashTag.findMany).toHaveBeenCalled();
+      // Should execute 3 SQL queries (one for each entity type: performer, studio, tag)
+      expect(prisma.$executeRaw).toHaveBeenCalledTimes(3);
     });
   });
 });

